### ISSUE TITLE
Improve stability propagation

### ISSUE THREAD

**Issue by sgflt** [ta4j/ta4j#1141]
**Is your feature request related to a problem? Please describe.**
```BaseStrategy``` should not be directly aware about stability. Stability is a function of used rules and indicators.
As user I have no clue about possible instabilities of used indicators and rules, therefore instability of strategy should be calculated at runtime from used rules.

**Describe the solution you'd like**
Drop unstableBars field from BaseStrategy and introduce isStableAt(int index) method to Rule.

F.E.:

```java
public class DownTrendRule extends AbstractRule {

  private static final int UNSTABLE_BARS = 5;

  private final ADXIndicator directionStrengthIndicator;
  private final Rule downtrendRule;


  public DownTrendRule(final BarSeries series) {
    this.directionStrengthIndicator = new ADXIndicator(series, UNSTABLE_BARS);

    final var plusDMIndicator = new PlusDMIndicator(series);
    final var minusDMIndicator = new MinusDMIndicator(series);
    this.downtrendRule = new UnderIndicatorRule(plusDMIndicator, minusDMIndicator);
  }


  @Override
  public boolean isSatisfied(final int index) {
    return this.directionStrengthIndicator.getValue(index).isGreaterThan(numOf(25))
           && this.downtrendRule.isSatisfied(index)
  }
}
```

There is no way how to propagate stability to strategy.

**Comment by sgflt** [comment]
Related question #1142

**Comment by sgflt** [comment]
If I consider folloowing implementation:

```java
public class DownTrendIndicator extends AbstractIndicator<Boolean> {

  private static final int UNSTABLE_BARS = 5;

  private final ADXIndicator directionStrengthIndicator;
  private final Rule downtrendRule;


  public DownTrendIndicator(final BarSeries series) {
    super(series);
    this.directionStrengthIndicator = new ADXIndicator(series, UNSTABLE_BARS);

    final var plusDMIndicator = new PlusDMIndicator(series);
    final var minusDMIndicator = new MinusDMIndicator(series);
    this.downtrendRule = new UnderIndicatorRule(plusDMIndicator, minusDMIndicator);
  }



  @Override
  public Boolean getValue(final int index) {
    return  this.directionStrengthIndicator.getValue(index).isGreaterThan(numOf(25))
            && this.downtrendRule.isSatisfied(index);
  }


  @Override
  public int getUnstableBars() {
    return UNSTABLE_BARS;
  }
}
```

Still, BaseStrategy have to be manually filled with unstableBars, which as user, I do not know where to get.

Related to #919

**Comment by sgflt** [comment]
If propagation through rules to strategy is not viable, then I am thinking about user friendly builder, which does the dirty work.

```java

StrategyBuilder.create()
   .withIndicators(List<Indicator>)    // Takes max unstable bars
   .withEntryRule(Rule)
   .withExitRule(Rule)
   .build() // calls new BaseStrategy(entry, exit, unstableBars)
   ;
```

**Comment by TheCookieLab** [comment]
Regarding your builder example, the Strategy should only need to know about its two Rules and not the Indicators they use.

An Indicator knows it's own unstable period. If we are returning NaN for any calculation within the unstable period then Rules don't need to know about the unstable period either as it can simply key off of NaN values. Typically Rules are comparing the results of one or more Indicators and if any of them are NaN it should know what to do.

**Comment by sgflt** [comment]
If NaN is used as unstable marker, it should be used consistently. But for example SMA returns estimates in unstable period. As contra argument, returning NaN inside unstable period cannot be distinguised from returning NaN within stable period that is caused by corrupted data or some zero division bug.

In PoC https://github.com/ta4j/ta4j/pull/1148 I have opted in for explicit isStable method to allow indicators return any value that is based on partial computation and that should not be used for decision. 

https://github.com/ta4j/ta4j/blob/2f176c9ebcc7bdf982ebd777e82c043d6146069f/ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/RunningTotalIndicator.java#L89-L92


Assuming Rule and Strategy have now stability property too and are aware of underlying indicators stability.

https://github.com/ta4j/ta4j/blob/71ca64e54404957821b2801daf5baeea4485b83c/ta4j-core/src/main/java/org/ta4j/core/Strategy.java#L79-L81

**Comment by TheCookieLab** [comment]
See https://github.com/ta4j/ta4j/issues/947

**Comment by sgflt** [comment]
https://github.com/ta4j/ta4j/issues/947#issuecomment-1561507718

This sums up concern about NaN. If NaN could be correct value within unstable period, each indicator have to handle NaN.

If we return normal numbers, indicators fixes automatically after some period of time without any NaN checking ifs.

```java
   // boilerplate
   var value = nestedIndicator.getValue();
   var value2 = nestedIndicator2.getValue();
   if (value.isNan() || value2.isNan()) {
      return NaN;
   }
```

Another example.
```java
  // sma is unusable 10 bars, 5 inner sma returns NaN, +5 NaN from outer sma
  // isStable approach is ready just within 5 bars, not in 10
  var sma = new SMAIndicator(new SMAIndicator(closePriceIndicator, 5), 5)
```

**Comment by sgflt** [comment]
Of course, but I am not sure about indexes. I think I have some difficulties so I actually I  have moved towards discrete simulation in  #1148 that goes only forward. It is completed from about 75 %. Could you have a look whether it is a way to go? 

At least  it is a show case what could be backported. I have used own cleaned code style so it is unmergable (unless consensual switching main branch to 1.0 version without merge) but it may be inspirative. Also I took look to some algorithms and reduced time complexity mostly from O(N^2) to O(N). Maybe much more in some classes.

**Comment by sgflt** [comment]
Uf, just started PoC of this change and it is going to be huge.

Dropping index in isStable causes cascade to Rule.isSatisfied. That leads to Indicator interface that have to drop indexing too. 

Maybe it would be easier to somehow switch branches to branch with 7 month of refactoring than reimplementing it again.

**Comment by sgflt** [comment]
Had to refresh what I did.

I have introduced IndicatorContext that allows me to collect all indicators data into prometheus metrics endpoint or store them within timescaledb.


IndicatorContext provides isStable() with simple implementation.
https://github.com/ta4j/ta4j/blob/a7c9bba7b3bcd5d867b2afef69ab716c2d68c4cb/ta4j-core/src/main/java/org/ta4j/core/indicators/IndicatorContext.java

Strategy then holds context and tests whether it is stable.
```java
  @Override
  public boolean isStable() {
    return this.indicatorContext.isStable();
  }
```
https://github.com/ta4j/ta4j/blob/a7c9bba7b3bcd5d867b2afef69ab716c2d68c4cb/ta4j-core/src/main/java/org/ta4j/core/backtest/BacktestStrategy.java

So many core breaking changes.
